libraries/AP_InertialSensor/AP_InertialSensor_Backend.cpp

```diff
@@ -174,6 +174,42 @@ void AP_InertialSensor_Backend::_publish_gyro(uint8_t instance, const Vector3f &
     _imu._delta_angle_acc_dt[instance] = 0;
 }
 
+/*
+  apply harmonic notch and low pass gyro filters
+ */
+void AP_InertialSensor_Backend::apply_gyro_filters(const uint8_t instance, const Vector3f &gyro)
+{
+    Vector3f gyro_filtered = gyro;
+
+    // apply the harmonic notch filters
+    for (auto &notch : _imu.harmonic_notches) {
+        if (!notch.params.enabled()) {
+            continue;
+        }
+#ifndef HAL_BUILD_AP_PERIPH
+        // by default we only run the expensive notch filters on the currently active IMU
+        if (!notch.params.hasOption(HarmonicNotchFilterParams::Options::EnableOnAllIMUs) &&
+            instance != AP::ahrs().get_primary_gyro_index()) {
+            continue;
+        }
+#endif
+        gyro_filtered = notch.filter[instance].apply(gyro_filtered);
+    }
+
+    // apply the low pass filter last to attentuate any notch induced noise
+    gyro_filtered = _imu._gyro_filter[instance].apply(gyro_filtered);
+
+    // if the filtering failed in any way then reset the filters and keep the old value
+    if (gyro_filtered.is_nan() || gyro_filtered.is_inf()) {
+        _imu._gyro_filter[instance].reset();
+        for (auto &notch : _imu.harmonic_notches) {
+            notch.filter[instance].reset();
```

Comment:
should check enabled here as well?

Reply:
resetting when not enabled will be faster than checking if it is enabled. The initialised check in HarmonicNotchFilter<T>::reset() will cause early return